packages/@o3r/eslint-plugin/src/rules/typescript/no-multiple-type-configuration-property/no-multiple-type-configuration-property.ts

```diff
@@ -0,0 +1,79 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import { createRule, isExtendingConfiguration } from '../../utils';
+
+export interface NoMultipleTypeConfigurationPropertyOption {
+  supportedInterfaceNames?: string[];
+}
+
+export default createRule<NoMultipleTypeConfigurationPropertyOption[], 'error' | 'suggestion'>({
+  name: 'no-multiple-type-configuration-property',
+  meta: {
+    hasSuggestions: true,
+    fixable: 'code',
+    type: 'problem',
+    docs: {
+      description: 'Ensures that the configuration property does not accept multiple type.',
+      recommended: 'error'
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          supportedInterfaceNames: {
+            type: 'array',
+            items: {
+              type: 'string'
+            },
+            default: ['NestedConfiguration', 'Configuration']
+          }
+        }
+      }
+    ],
+    messages: {
+      error: 'Configuration could not be the union of multiple type',
+      suggestion: 'Replace {{currentValue}} by {{recommendedValue}}'
+    }
+  },
+  defaultOptions: [],
+  create: (context) => {
+    const supportedInterfaceNames = context.options.reduce((acc: string[], option) => acc.concat(option.supportedInterfaceNames || []), []);
+
+    const rule = (node: TSESTree.TSUnionType | TSESTree.TSIntersectionType) => {
+      const interfaceDeclNode = node.parent?.parent?.parent?.parent;
+      if (!isExtendingConfiguration(interfaceDeclNode, supportedInterfaceNames)) {
+        return; // Not in a configuration interface
+      }
+
+      if (
+        node.types.every((type) => type.type === TSESTree.AST_NODE_TYPES.TSLiteralType && type.literal.type === TSESTree.AST_NODE_TYPES.Literal)
+        && [...(new Set((node.types as TSESTree.TSLiteralType[]).map((literalType) => typeof (literalType.literal as TSESTree.Literal).value)))].length === 1
+      ) {
+        return; // Only the same literal type
+      }
+
+      const text = context.getSourceCode().getText(node);
+      const separatorRegExp = /\s*(\||&)\s*/;
```

Comment:
Could be defined outside of the function (lets reduce the cost of the linter evaluation as much as possible for our poor IDEs)